# **Data Cleaning & Feature Engineering**

## Objectives

* Clean and prepare the dataset
* Split the dataset into respective train and test sets for Outcome of Diabetic and Non-Diabetic
* Impute zero values with the median
* Standardise the data ready for training the ML model
* Select the feature engineering method for use in the ML pipeline using the train test datasets

## Inputs

* outputs/datasets/collection/diabetes.csv

## Outputs

* Cleaned dataset will be generated for Train and Test sets and will be exported to outputs/datasets/cleaned folder
    * outputs/datasets/cleaned/x_train_cleaned.csv
    * outputs/datasets/cleaned/y_train_cleaned.csv
    * outputs/datasets/cleaned/x_test_cleaned.csv
    * outputs/datasets/cleaned/y_test_cleaned.csv

## Additional Comments

* This Notebook falls under the CRISP-DM of Data Preperation.
* The conclusions of this notebook is that we have a Cleaned Data Pipeline ready for the model and have split the dataset into train and test sets.
* Due to the overlap between the data cleaning and feature engineering portions of the pipeline, it was decided that one notebook was best used to incorporate both the Data Cleaning and the Feature engineering processes.


---

# Change working directory

* As the notebooks are stored in the subfolder 'jupyter_notebooks' we therefore, when running the notebook in the editor, need to change the working directory.

We need to change the working directory from its current folder to its parent folder
* We access the current directory with os.getcwd()

In [1]:
import os
current_dir = os.getcwd()
current_dir

'/workspace/pp5-diabetes-prediction/jupyter_notebooks'

We want to make the parent of the current directory the new current directory
* os.path.dirname() gets the parent directory
* os.chir() defines the new current directory

In [2]:
os.chdir(os.path.dirname(current_dir))
print("You set a new current directory")

You set a new current directory


Confirm the new current directory

In [ ]:
current_dir = os.getcwd()
current_dir

# Importing the Libraries

* Here we will import the dependencies used during the Correlation Study phase.

In [3]:
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt
%matplotlib inline

from sklearn.preprocessing import StandardScaler
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score

# Loading Collected Dataset

* We will begin by loading the diabetes dataset.

* A Pandas dataframe is declared using the diabetes dataset using `read_csv()`

* The first fifteen rows will be displayed using `head()` to get a broader view of the data that will need cleaning

In [4]:
import pandas as pd
df = pd.read_csv(f"outputs/datasets/collection/diabetes.csv")
df.head(15)


Pregnancies  Glucose  BloodPressure  SkinThickness  Insulin   BMI  \
0             6      148             72             35        0  33.6   
1             1       85             66             29        0  26.6   
2             8      183             64              0        0  23.3   
3             1       89             66             23       94  28.1   
4             0      137             40             35      168  43.1   
5             5      116             74              0        0  25.6   
6             3       78             50             32       88  31.0   
7            10      115              0              0        0  35.3   
8             2      197             70             45      543  30.5   
9             8      125             96              0        0   0.0   
10            4      110             92              0        0  37.6   
11           10      168             74              0        0  38.0   
12           10      139             80              0        0  27.1   
13            1      189             60             23      846  30.1   
14            5      166             72             19      175  25.8   

    DiabetesPedigreeFunction  Age  Outcome  
0                      0.627   50        1  
1                      0.351   31        0  
2                      0.672   32        1  
3                      0.167   21        0  
4                      2.288   33        1  
5                      0.201   30        0  
6                      0.248   26        1  
7                      0.134   29        0  
8                      0.158   53        1  
9                      0.232   54        1  
10                     0.191   30        0  
11                     0.537   34        1  
12                     1.441   57        0  
13                     0.398   59        1  
14                     0.587   51        1

---

# Zero Value Data

* We identified previously that some variables had some abnormalities with zero values, this is likely due to missing values being represented as zeros in the original source data.

* These zero values will either need to be removed from the dataset or replaced and added by imputing with a median value. As the dataset is already small, imputing will be opted for in this stage of cleaning the data.

* External code was taken from [towardsdatascience](https://towardsdatascience.com/pima-indian-diabetes-prediction-7573698bd5fe) for guidance with replacing zero values and was then customised to be appropriate for our dataset.

In [5]:
# BloodPressure

df1 = df.loc[df['Outcome'] == 1]
df2 = df.loc[df['Outcome'] == 0]
df1 = df1.replace({'BloodPressure':0}, np.median(df1['BloodPressure']))
df2 = df2.replace({'BloodPressure':0}, np.median(df2['BloodPressure']))
dataframe = [df1, df2]
df = pd.concat(dataframe)

* Zero values will now be replaced for 1 (Diabetic) with a median of 1 and this will also be true for 0 (Non-Diabetic) when it comes to the Outcome variable using NumPy's median function.

* We need to apply the same code for each variable where there is an abnormal occurance.

In [6]:
# BMI

df1 = df.loc[df['Outcome'] == 1]
df2 = df.loc[df['Outcome'] == 0]
df1 = df1.replace({'BMI':0}, np.median(df1['BMI']))
df2 = df2.replace({'BMI':0}, np.median(df2['BMI']))
dataframe = [df1, df2]
df = pd.concat(dataframe)

In [7]:
# Insulin

df1 = df.loc[df['Outcome'] == 1]
df2 = df.loc[df['Outcome'] == 0]
df1 = df1.replace({'Insulin':0}, np.median(df1['Insulin']))
df2 = df2.replace({'Insulin':0}, np.median(df2['Insulin']))
dataframe = [df1, df2]
df = pd.concat(dataframe)

In [8]:
# SkinThickness

df1 = df.loc[df['Outcome'] == 1]
df2 = df.loc[df['Outcome'] == 0]
df1 = df1.replace({'SkinThickness':0}, np.median(df1['SkinThickness']))
df2 = df2.replace({'SkinThickness':0}, np.median(df2['SkinThickness']))
dataframe = [df1, df2]
df = pd.concat(dataframe)

In [9]:
# Glucose
df1 = df.loc[df['Outcome'] == 1]
df2 = df.loc[df['Outcome'] == 0]
df1 = df1.replace({'Glucose':0}, np.median(df1['Glucose']))
df2 = df2.replace({'Glucose':0}, np.median(df2['Glucose']))
dataframe = [df1, df2]
df = pd.concat(dataframe)

* We can test that the above changes have been made by running the function `head()` to get a list of the first fifteen rows again to see what the new median data will be.

In [10]:
df.head(15)

Pregnancies  Glucose  BloodPressure  SkinThickness  Insulin    BMI  \
0             6      148             72             35        0  33.60   
2             8      183             64             27        0  23.30   
4             0      137             40             35      168  43.10   
6             3       78             50             32       88  31.00   
8             2      197             70             45      543  30.50   
9             8      125             96             27        0  34.25   
11           10      168             74             27        0  38.00   
13            1      189             60             23      846  30.10   
14            5      166             72             19      175  25.80   
15            7      100             74             27        0  30.00   
16            0      118             84             47      230  45.80   
17            7      107             74             27        0  29.60   
19            1      115             70             30       96  34.60   
22            7      196             90             27        0  39.80   
23            9      119             80             35        0  29.00   

    DiabetesPedigreeFunction  Age  Outcome  
0                      0.627   50        1  
2                      0.672   32        1  
4                      2.288   33        1  
6                      0.248   26        1  
8                      0.158   53        1  
9                      0.232   54        1  
11                     0.537   34        1  
13                     0.398   59        1  
14                     0.587   51        1  
15                     0.484   32        1  
16                     0.551   31        1  
17                     0.254   31        1  
19                     0.529   32        1  
22                     0.451   41        1  
23                     0.263   29        1

* After cleaning the data we can see that Insulin still has zero values for the median for Diabetics (Outcome 1). From this we hypothesise that Insulin for Diabetics is lower than that of non-diabetics.

* Next we want to group the diabetes dataset by mean for the Outcome as this is what the machine learning algorithm will see.

In [11]:
df.groupby('Outcome').mean()

Pregnancies     Glucose  BloodPressure  SkinThickness     Insulin  \
Outcome                                                                      
0           3.298000  110.622000      70.844000      25.502000   87.200000   
1           4.865672  142.302239      75.242537      31.029851  100.335821   

               BMI  DiabetesPedigreeFunction        Age  
Outcome                                                  
0        30.845100                  0.429734  31.190000  
1        35.398134                  0.550500  37.067164

* We then want to separate the variables data from the Outcome label by dropping the Outcome column using the `drop()` function.

In [12]:
x = df.drop(columns = 'Outcome', axis=1)
y = df['Outcome']

In [13]:
print(x)

     Pregnancies  Glucose  BloodPressure  SkinThickness  Insulin   BMI  \
0              6      148             72             35        0  33.6   
2              8      183             64             27        0  23.3   
4              0      137             40             35      168  43.1   
6              3       78             50             32       88  31.0   
8              2      197             70             45      543  30.5   
..           ...      ...            ...            ...      ...   ...   
762            9       89             62             21       39  22.5   
763           10      101             76             48      180  32.9   
764            2      122             70             27       39  36.8   
765            5      121             72             23      112  26.2   
767            1       93             70             31       39  30.4   

     DiabetesPedigreeFunction  Age  
0                       0.627   50  
2                       0.672   32  


* As we can see we have successfully separated the Outcome column as it is no longer showing with the rest of the dataset when printing 'x'.

In [14]:
print(y)

0      1
2      1
4      1
6      1
8      1
      ..
762    0
763    0
764    0
765    0
767    0
Name: Outcome, Length: 768, dtype: int64


* Next we want to standardise the data to make it easier for the machine learning model to predict the outcome.

## Data Standardisation

In [17]:
df_scaler = StandardScaler()

* We use `fit()` and `transform()` to take the training data as an argument and compute the mean and standard deviation to be used for further scaling.

In [18]:
df_scaler.fit(x)
stnd_data = df_scaler.transform(x)
print(stnd_data)

[[ 0.63994726  0.86462486 -0.0313235  ...  0.16958256  0.46849198
   1.4259954 ]
 [ 1.23388019  2.01426457 -0.69266918 ... -1.3283415   0.60439732
  -0.10558415]
 [-1.14185152  0.50330953 -2.67670622 ...  1.551163    5.4849091
  -0.0204964 ]
 ...
 [-0.54791859  0.01060679 -0.19665992 ...  0.63495703 -0.39828208
  -0.53102292]
 [ 0.3429808  -0.02224005 -0.0313235  ... -0.90659589 -0.68519336
  -0.27575966]
 [-0.84488505 -0.94195182 -0.19665992 ... -0.2957919  -0.47378505
  -0.87137393]]


* We will use the variables 'x' and 'y' for training our machine learning model. 'x' represents the data and 'y' represents the model.

In [19]:
x = stnd_data
y = df['Outcome']
print(x)
print(y)

[[ 0.63994726  0.86462486 -0.0313235  ...  0.16958256  0.46849198
   1.4259954 ]
 [ 1.23388019  2.01426457 -0.69266918 ... -1.3283415   0.60439732
  -0.10558415]
 [-1.14185152  0.50330953 -2.67670622 ...  1.551163    5.4849091
  -0.0204964 ]
 ...
 [-0.54791859  0.01060679 -0.19665992 ...  0.63495703 -0.39828208
  -0.53102292]
 [ 0.3429808  -0.02224005 -0.0313235  ... -0.90659589 -0.68519336
  -0.27575966]
 [-0.84488505 -0.94195182 -0.19665992 ... -0.2957919  -0.47378505
  -0.87137393]]
0      1
2      1
4      1
6      1
8      1
      ..
762    0
763    0
764    0
765    0
767    0
Name: Outcome, Length: 768, dtype: int64


* As we can see above, scikit-learn strips the column labels from our data when converting to a numpy ndarray for standardising the data. Whilst this is fine at this current stage, in case we need to manipulate the data further down the line, this can make things confusing. We will need to add the column labels back on for ease of interpretation.

In [34]:
x_col = df.columns[0:-1]
print(x_col)


Index(['Pregnancies', 'Glucose', 'BloodPressure', 'SkinThickness', 'Insulin',
       'BMI', 'DiabetesPedigreeFunction', 'Age'],
      dtype='object')


* As you can see above, these are the original column labels that we need to include back into our dataset. We will achieve this using the following code where the solution was found for guidance from [Stackoverflow](https://stackoverflow.com/questions/29586323/how-to-retain-column-headers-of-data-frame-after-pre-processing-in-scikit-learn)

In [32]:
x_df = pd.DataFrame(x, columns = x_col)
print(x_df)

     Pregnancies   Glucose  BloodPressure  SkinThickness   Insulin       BMI  \
0       0.639947  0.864625      -0.031324       0.812528 -0.849452  0.169583   
1       1.233880  2.014265      -0.692669      -0.046266 -0.849452 -1.328341   
2      -1.141852  0.503310      -2.676706       0.812528  0.705374  1.551163   
3      -0.250952 -1.434655      -1.850024       0.490480 -0.035019 -0.208534   
4      -0.547919  2.474120      -0.196660       1.886020  4.175966 -0.281249   
..           ...       ...            ...            ...       ...       ...   
763     1.530847 -1.073339      -0.858006      -0.690362 -0.488510 -1.444685   
764     1.827813 -0.679177       0.299349       2.208068  0.816433  0.067782   
765    -0.547919  0.010607      -0.196660      -0.046266 -0.488510  0.634957   
766     0.342981 -0.022240      -0.031324      -0.475663  0.187099 -0.906596   
767    -0.844885 -0.941952      -0.196660       0.383131 -0.488510 -0.295792   

     DiabetesPedigreeFunction       Age

* We have now converted the x matrix back to a pandas dataframe and included the original column labels ready to be split into the respective train and test sets.

---

## Train Test Split

* Next we will need to split the cleaned data into the train and test sets.
* The x variable data will be split into two arrays which are x_train and x_test.
* The test size will be 10% of the dataset and the train size will be 90% of the dataset. As the dataset is small, we want to maximise the amount of trained data to better improve the model.

In [38]:
x_train, x_test, y_train, y_test = train_test_split(x_df,y, test_size = 0.1, stratify=y, random_state=2)
print(f"Total Records: {x.shape} \nx_train set: {x_train.shape} \nx_test set: {x_test.shape}")

Total Records: (768, 8) 
x_train set: (691, 8) 
x_test set: (77, 8)


* From this, 768 is the number of records in our dataset with 691 records being used for training data leaving 77 records to be used for the test data.

* Next we need to move onto Feature Engineering and train the model.

---

# Pushing the clean files to Repository

* If you do not need to push files to Repo, you may replace this section with "Conclusions and Next Steps" and state your conclusions and next steps.

In [21]:
try:
    # Creates cleaned folder in the outputs directory
    os.makedirs(name='outputs/datasets/cleaned')
except Exception as e:
    print(e)

[Errno 17] File exists: 'outputs/datasets/cleaned'


### Train Dataset

In [39]:
pd.DataFrame(x_train).to_csv("outputs/datasets/cleaned/x_train_cleaned.csv", index=False)

In [49]:
pd.DataFrame(y_train).to_csv("outputs/datasets/cleaned/y_train_cleaned.csv", index=False)

### Test Dataset

In [50]:
pd.DataFrame(x_test).to_csv("outputs/datasets/cleaned/x_test_cleaned.csv", index=False)

In [51]:
pd.DataFrame(y_test).to_csv("outputs/datasets/cleaned/y_test_cleaned.csv", index=False)

---

# Loading Train/Test Dataset

#### Train Set

In [52]:
import pandas as pd

x_train_path = "outputs/datasets/cleaned/x_train_cleaned.csv"
x_train = pd.read_csv(x_train_path)

x_train.head(15)

Pregnancies   Glucose  BloodPressure  SkinThickness   Insulin       BMI  \
0      0.342981  0.569003      -0.692669       0.812528  0.446236 -0.557565   
1     -0.547919 -0.482096       0.134013       0.275781  0.076040  0.169583   
2     -0.250952 -0.449249      -0.858006      -0.368314 -0.488510 -0.935682   
3     -0.844885 -0.219321      -0.196660       0.275781  0.039020  0.315012   
4     -0.844885 -1.139033       0.464686      -0.046266 -0.553294  0.315012   
5      0.639947 -0.941952      -1.850024       0.275781 -0.257137 -0.543022   
6     -0.547919 -0.712024      -0.527333      -0.797711 -0.016509  0.067782   
7      0.046014 -0.810564      -1.023342      -0.475663 -0.488510 -0.615737   
8     -1.141852 -0.646330      -1.684688      -0.690362 -0.488510 -1.066568   
9     -1.141852  0.043454       1.291368       1.027226 -0.488510  0.402270   
10     1.233880  0.371922      -0.031324      -0.046266 -0.849452  0.067782   
11     1.827813  1.291634      -0.361996      -0.475663  0.372197 -1.008397   
12     1.530847  1.127400       1.126031      -0.046266 -0.849452 -1.110197   
13    -0.844885  0.207688       0.795359      -1.119759  0.844197 -0.717538   
14    -0.547919 -0.646330       1.126031       0.919877  0.261138  1.900194   

    DiabetesPedigreeFunction       Age  
0                  -0.183854 -0.616111  
1                  -0.204994 -0.871374  
2                  -0.751636 -0.701198  
3                   0.172520 -0.105584  
4                  -1.120090 -0.956462  
5                  -0.349960 -0.871374  
6                   1.193320 -0.445935  
7                  -0.087210 -0.956462  
8                  -1.189553 -1.041549  
9                  -0.830159 -0.360847  
10                 -0.609690  0.490030  
11                 -0.440564  1.170732  
12                 -0.730495  1.681259  
13                 -1.077809 -0.956462  
14                 -1.041567 -0.871374

#### Test Set

In [53]:
x_test_path = "outputs/datasets/cleaned/x_test_cleaned.csv"
x_test = pd.read_csv(x_test_path)

x_test.head(15)

Pregnancies   Glucose  BloodPressure  SkinThickness   Insulin       BMI  \
0      0.342981  0.831778       0.464686      -0.690362 -0.488510  0.184126   
1     -0.844885 -0.876258       0.795359      -0.260965  0.816433  0.373184   
2     -0.250952 -0.350709      -0.858006      -0.690362 -0.488510 -1.430142   
3      0.639947 -0.416402      -1.023342      -0.046266 -0.488510 -1.081111   
4     -0.250952 -0.219321      -0.527333       1.241924  0.446236  0.824015   
5     -0.844885 -1.040492      -0.858006      -1.012410 -0.303412 -1.066568   
6     -0.844885  0.798931      -1.354015      -0.690362 -0.488510 -0.397593   
7      1.233880 -0.055087       0.464686      -0.690362 -0.488510 -1.081111   
8     -1.141852  0.503310       0.960695      -0.046266 -0.488510 -0.746623   
9      0.046014  0.240535      -1.023342      -1.656505  1.288433 -0.717538   
10    -0.844885 -1.106186       0.464686       0.168432 -0.146078 -0.063105   
11     2.124780 -0.613483      -0.361996       1.349274 -0.488510  2.001995   
12    -0.547919 -0.449249      -0.858006       0.490480 -0.331176 -1.052025   
13     0.342981  2.145652       0.299349      -0.046266  1.066315  1.623878   
14    -0.547919  0.798931      -0.196660       1.134575  2.482317 -0.644823   

    DiabetesPedigreeFunction       Age  
0                  -0.766737  2.702312  
1                  -0.721435  0.830381  
2                  -0.996266 -1.041549  
3                  -0.802978 -0.531023  
4                  -0.972105 -0.445935  
5                   2.404388 -0.701198  
6                   0.278225 -0.360847  
7                  -0.189894  2.617224  
8                  -0.727475  2.191785  
9                   0.166480 -0.190672  
10                 -0.322779 -0.360847  
11                 -1.044587  0.745293  
12                 -1.038547 -1.041549  
13                  1.697680  1.681259  
14                 -0.407342 -0.360847

---